students/MicahBraun/Lesson 05/mailroom_3.py

```diff
@@ -0,0 +1,297 @@
+    directions = "Select one of the three options:"  # directions to user for menu
+    optA = "[A] Send a Thank you"  # option(s) for menu
+    optB = "[B] Create a Report"
+    optC = "[C] Create Letters for All Donors"
+    optD = "[D] Quit"
+    whole_menu_header = header + '\n' + title + '\n' + header  # header strung together as one uni
+
+    while True:
+        print(whole_menu_header)  # display header
+        print(directions.center(110, ' '), '\n')  # center directions with header display
+        print(optA + optB.center(37, ' ') + optC.center(36, ' ') + optD.center(25, ' ') + '\n')
+
+        switch_function = {
+            'A': thankyou,  # dictionary holding menu options
+            'B': createreport,
+            'C': writeletters,
+            'D': quit_program,
+        }
+        try:
+            usr_sel = (input("Menu Selection: ")).upper()  # user input for menu
+            switch_function.get(usr_sel)()  # get user entry from dict
+
+        except Exception as e:
+            print("Invalid Entry! Only enter A, B, C, or D. \n", e)  # display error on any other entry
+        continue  # continue back to start of menu
+
+
+def thankyou():
+    """Function enables user to add to dictionary data (adding new donors and dontation amounts as well as
+    number of donations and average donations for new keys, and also updates existing keys."""
+    print("Leaving menu...\n")
+    while True:
+        print("Enter the name of the person you are writing to (or enter 'list' to see a list of names or Q to quit) ")
+        fname_prompt = input("First Name: ").strip().capitalize()  # first name variable (strip any spaces)
+        if fname_prompt.upper() == "Q":  # if Q, then quit to menu()
+            menu()
+        elif fname_prompt.lower() == "list":
+            displaylist()
+        else:
+            lname_prompt = input("Last Name: ").strip().capitalize()  # last name variable
+            if lname_prompt.upper() == "Q":  # if Q, back to menu
+                menu()
+            elif lname_prompt.lower() == "list":  # if list, then list display
+                displaylist()
+            else:
+                key = fname_prompt + " " + lname_prompt
+                if key in donor_dict.keys():
+                    existing_donor = input(
+                        "That value is already in the list! Do you want to proceed with that selection? (Y/N): ")
+                    if existing_donor.upper() == "Y":  # if user proceeds, print display
+                        donation_amt = int(input("Enter in the donation amount from Donor {0}: $".format(key)))
+                        if str(donation_amt).upper() == "Q":
+                            menu()
+                        else:
+                            print(
+                                '{0} has donated ${1:,.2f}'.format(key, donation_amt))  # display name and donation amt
+                            int(donation_amt)
```

Comment:
This would be a good place for a try/except on the int conversion to catch non-numerical input